AppKitNavigation - Navigation

Mx-Iris opened the pull request:
This PR provides AppKit navigation with present, sheet, modal, modal session

stephencelis commented:
@Mx-Iris The other branch was merged so we should be able to review this one as soon as the conflicts are resolved. Can you ping us when it's ready?

Mx-Iris commented:
@stephencelis The conflict has been resolved and is ready for review

stephencelis commented on Sources/AppKitNavigation/Internal/AssociatedKeys.swift:
```diff
@@ -0,0 +1,36 @@
+struct AssociatedKeys {
+    var keys: [AnyHashableMetatype: UnsafeMutableRawPointer] = [:]
+
+    mutating func key<T>(of type: T.Type) -> UnsafeMutableRawPointer {
```
From what I can tell this is a helper that sets up an associated object. Is it doing anything else? If not, I'm inclined to be consistent with the pattern used in UIKitNavigation, where we define dedicated local keys where needed instead of leveraging a dynamic helper. How's that sound?

Mx-Iris replied:
Again because there is more than one type of object that can navigate, this is mainly used for Observer generic types like Sheet, the classes that can execute it are NSWindow, NSSavePanel, NSAlert, NSPrintPanel.... They are not necessarily of the same type, and each class has a different way of executing a sheet.

stephencelis replied:
Because associated keys are local to each object, does the type need to be encoded into the operation? Wouldn't it work the same to use the same static key instead?

Mx-Iris replied:
```
private func sheetObserver<Content: SheetContent>() -> SheetObserver<Self, Content> {
        if let observer = objc_getAssociatedObject(self, sheetObserverKeys.key(of: Content.self)) as? SheetObserver<Self, Content> {
            return observer
        } else {
            let observer = SheetObserver<Self, Content>(owner: self)
            objc_setAssociatedObject(self, sheetObserverKeys.key(of: Content.self), observer, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
            return observer
        }
    }
```
Because of the one-to-many relationship, a class that executes a sheet may have more than one observer (this class is mainly used to store the object that has been executed sheet), the object that is executed by the sheet is indeterminate, it may be NSAlert or NSWindow. Currently, storing the object that is executed by the sheet is mainly to perform cleanup work, if you have a If you have a better idea, let me know! 😄

stephencelis commented on Sources/AppKitNavigation/Navigation/Sheet.swift:
```diff
@@ -0,0 +1,132 @@
+    }
+}
+
+extension NSWindow {
+    func endSheeted() {
+        guard sheetParent != nil else {
+            return
+        }
+        sheetParent?.endSheet(self)
+    }
+}
+
+extension Navigated where Content: SheetContent {
+    func clearup() {
+        content?.currentWindow?.endSheeted()
+    }
+}
+
```
Can simplify this I think:

```suggestion
extension Navigated where Content: SheetContent {
    func clearup() {
        guard let window = content?.currentWindow else { return }
        window.sheetParent?.endSheet(window)
    }
}

```

stephencelis commented on Sources/AppKitNavigationShim/include/shim.h:
```diff
@@ -0,0 +1,19 @@
+
+@interface NSViewController (AppKitNavigation)
+
+@property BOOL hasViewAppeared;
+@property (nullable) void (^ onDismiss)();
+@property NSArray<void (^)()> *onViewAppear;
```
We've had to prefix these in UIKitNavigation to avoid collisions with external code.

```suggestion
@property BOOL _AppKitNavigation_hasViewAppeared;
@property (nullable) void (^ _AppKitNavigation_onDismiss)();
@property NSArray<void (^)()> *_AppKitNavigation_onViewAppear;
```

stephencelis commented on Sources/AppKitNavigation/Navigation/ModalSessionContent.swift:
```diff
@@ -0,0 +1,22 @@
+
+@MainActor
+public protocol ModalSessionContent: ModalContent {
+    func appKitNavigationBeginModalSession() -> NSApplication.ModalSession
```
Do you think it's possible to avoid these protocols? In UIKitNavigation we got by using the classes themselves, but maybe I'm missing something?

Mx-Iris replied:
The controls that can perform ModalSession are not only NSWindow, but also NSAlert, NSSavePanel, etc. The same applies to other Sheet protocols.

maximkrouk replied:
There are a few comments about public protocols that might be a little excessive for the public API, if they greatly simplify things internally I think there might be an option to hide them using `@_spi` 🤔

Mx-Iris replied:
@maximkrouk I tried it and found that it doesn't work. If you use this protocol for spi, all methods that use this protocol need to be spi-compatible.

stephencelis commented:
@Mx-Iris Thanks for continuing to contribute :smile: I did a quick initial review and commented on things that stood out to me. I think the main thing we'd like to aim for is consistency with other parts of the library, _e.g._ UIKitNavigation. In particular:

- Let's stick with the existing pattern for associated objects
- Let's try to avoid introducing any public APIs (protocols, methods, etc.) that are only meant to be used internally.
- If possible, let's try to avoid introducing protocols to abstract concepts. And if there are good reasons to introduce protocols, let's discuss why.
- Let's also try to avoid helpers unless they're saving a lot of extra work.

Let us know what you think!

Mx-Iris commented on Sources/AppKitNavigation/Navigation/ModalSessionContent.swift:
```diff
@@ -0,0 +1,22 @@
+
+extension NSWindow: ModalSessionContent {
+
+    public func appKitNavigationBeginModalSession() -> NSApplication.ModalSession {
+        __appKitNavigationBeginModalSession()
+    }
+    
+    @objc func __appKitNavigationBeginModalSession() -> NSApplication.ModalSession {
+        let modalSession = NSApplication.shared.beginModalSession(for: self)
+        return modalSession
+    }
```
Because the Modal way of NSWindow is different from the Modal way of NSSavePanel, but NSSavePanel inherits from NSWindow, Swift protocol extensions can't override the methods in the subclasses, this one is not used yet, and it can be deleted for the time being.

Mx-Iris commented on Sources/AppKitNavigation/Navigation/NavigationContent.swift:
```diff
@@ -0,0 +1,38 @@
+import Foundation
+
+@MainActor
+public protocol NavigationContent: AnyObject {
```
This protocol abstracts a lot of navigation methods. UIKit's navigation is relatively homogeneous, but AppKit's navigation is messy and complex, and if you don't pull it out there will be a lot of duplicate code.

Mx-Iris commented:
@stephencelis Thanks for the review, I replied to some of the issues you mentioned because AppKit is very different from UIKit, it's very cluttered and can't be viewed in the same way as UIKit.

stephencelis commented:
@Mx-Iris Thanks again for your patience! @mbrandonw and I took another look today and we think it's a fantastic start but we're not yet comfortable enough with the code to bring it in and maintain it as is. We think there are opportunities to remove some of the public and internal abstractions and expose helpers that mirror the platform's built-in helpers in a similar way to the UIKitNavigation module.

We think there are a couple paths forward depending on your interest:

1. We could consider this branch a really great reference for @mbrandonw and I to look at when we try to reframe the solution in a style that makes sense to us. We're not 100% sure when we'll have the spare cycles for such a task, but hopefully in the coming months.

2. We could maybe narrow this PR's focus even further so that we have a minimal amount of code to review and push in a direction that makes sense to us as project maintainers. Right now the PR introduces a lot of different navigation styles that are each implemented in their own chunk of code: presentation, navigation, modals, and sheets. If you're interested in continuing to work on this project and refine things, we think it'd be good to pick the simplest of all these forms of navigation and open a dedicated PR that tackles just that form. It looks like the "presentation" APIs might be the simplest or at least closest in analog to a lot of the UIKitNavigation APIs. Does that sound right?

Do you have a preference here? We don't want to keep pushing back on your work if it feels annoying or nitpicky, but are happy to continue working with you if you're up for it. Otherwise, we hope you're OK with us taking on the next steps of this PR when we have a chance. Thoughts?

Mx-Iris commented:
@stephencelis Hey, Discussing the navigation separately indeed makes things clearer. Since I've been quite busy with work recently, I may not have much time to deal with these matters. Should we leave it here for now? I'll come back to finish it when I have time.

stephencelis commented:
@Mx-Iris Yup no rush at all! Thanks for the update. If you want to break things down to a smaller PR when you have time that would be helpful and instructive, and if Brandon and I have time we may push some commits ourselves :smile:

Mx-Iris commented:
@stephencelis I've set all protocols to internal to only expose explicit types externally, but this would be a lot of boilerplate code and does reduce complexity a lot, if you can, follow up by introducing macros to solve the boilerplate code issue, see the extension in the Modal navigation section to see if it addresses your concerns!

Mx-Iris commented:
@stephencelis, @mbrandonw Hey, are you free to do a review?
